### PR TITLE
Use leak sanitizer instead of internal mdebug to check for memory leaks

### PR DESCRIPTION
**Pull request by levitte**
The leak sanitizer gives better reports (complete stack traces) and
works as a wrapper around the application instead of relying on
cooperative enabling and disabling calls (which are too easy to get
unbalanced).

Related to #8322

### CODE REVIEW

**Comment by levitte** [comment]
This also replaces #8323

**Comment by mspncp** [comment]
fly-by-comment:  typo in commit message title (and GitHub title): 'sasnitizer'

**Comment by richsalz** [comment]
Should https://github.com/openssl/tools/blob/master/review-tools/opensslbuild#L36 also be changed?

**Comment by levitte** [comment]
> fly-by-comment: typo in commit message title (and GitHub title): 'sasnitizer'

Fixed

I also changed the way it's enabled, as sufficiently old clang doesn't understand `-fsanitize=leak`, but the address sanitizer has understood to enable the leak sanitizer internally for longer.

**Comment by levitte** [comment]
> Should https://github.com/openssl/tools/blob/master/review-tools/opensslbuild#L36 also be changed?

I think it should.  The leak sanitizer gives far better output than our mdebug ever will.

**Comment by mattcaswell** [comment]
Travis failure looks relevant:

````
==34177==AddressSanitizer: detect_leaks is not supported on this platform.
````

**Comment by levitte** [comment]
Damn...

**Comment by t-j-h** [comment]
Drop that platform out for this check.
It is possible to build a mingw cross compiler with asan support - but it isn't necessary to get the testing we need here for most linux-like platforms.

**Comment by levitte** [comment]
> Drop that platform out for this check.

I *am* worried about corner cases for code paths that only happen on platforms....  on the other hand, since this approach is clang/gcc centric already, maybe I worry too much.

**Comment by levitte** [comment]
I'll rework this to avoid mingw...  that will require some major surgery of the build matrix

**Comment by richsalz** [comment]
Are you still intending to drop the internal memory-leak tooling?  I totally support that in favor of ASAN.  ASAN is what everyone is doing. :)

**Comment by levitte** [comment]
BTW, I have *no* idea why mingw was mentioned at all here.  The Travis failure happens on MacOS X

**Comment by levitte** [comment]
The MacOS problem has been complained about here: https://github.com/google/sanitizers/issues/1026
Apparently, it comes down to clang...

**Comment by richsalz** [comment]
When you remove the mdebug stuff, please also remove all the fault-insertion stuff in crypto/mem.c (shouldfail(), md_trace_fd, etc)

**Comment by mattcaswell** [comment]
I think we should just disable this on MacOS.

**Comment by levitte** [comment]
I solved the MacOS case by not specifying `ASAN_OPTIONS="detect_leaks=1"` in .travis.yml.  Leak detection is on by default in modern enough ASAN, I hope that makes it all reliable enough.

**Comment by mspncp** [comment]
Sorry, I couldn’t resist any longer 😉.

**Comment by levitte** [comment]
That's all right. Thanks

**Comment by levitte** [comment]
I'm pondering if I should add a test program that intentionally leaks memory, for demonstration purposes...

**Review by paulidale — APPROVED** [review]
I'm inclined to think yes to the deliberately leaking test program.
We'd quickly find out if leak detection ever stopped working.

**Comment by richsalz** [comment]
we have aborttest and memleaktest

**Comment by levitte** [comment]
I just modified `test/memleaktest.c` to work with the sanitizer.

**Review by paulidale — DISMISSED** [review]

**Comment by levitte** [comment]
> Travis failure looks relevant

Yup

**Comment by levitte** [comment]
The appveyor failure doesn't seem relevant here

**Comment by levitte** [comment]
CIs are happy.  Is @openssl happy?

**Review by paulidale — APPROVED** [review]
I'm okay with this.

File: test/recipes/90-test_memleak.t [hunk 1]
```diff
@@ -10,6 +10,10 @@
 use OpenSSL::Test;
 
 setup("test_memleak");
+
+plan skip_all => "MacOS currently doesn't support leak sanitizer"
```
Comment (richsalz) [inline test/recipes/90-test_memleak.t]: Is isn't? https://clang.llvm.org/docs/LeakSanitizer.html

Comment (richsalz) [reply test/recipes/90-test_memleak.t]: I guess it depends on which version of clang/llvm you have :(

**Comment by richsalz** [comment]
are you going to remove the current mdebug cruft?

File: test/recipes/90-test_memleak.t (hunk 1 above)
Comment (levitte) [reply test/recipes/90-test_memleak.t]: Yup.

**Comment by levitte** [comment]
> are you going to remove the current mdebug cruft?

That's my intention, yes.

**Comment by levitte** [comment]
... or... wanna take that on? I wouldn't mind. You do need to retain the public API, though...

**Comment by richsalz** [comment]
I will make a PR to remove the existing mem-debug stuff.  I had wanted to remove the malloc-fail stuff as well, but it seems there's interest in this now.  (I forget which comment in which PR)

**Comment by richsalz** [comment]
PR #10572

**Comment by levitte** [comment]
Merged.

22c22369bc09988eb8c6fdf4f543581a78fd3eb1 Use leak sanitizer instead of internal mdebug to check for memory leaks
ea7a952c8aab33d0bb0a2bbd210305d722a5702b test/memleaktest.c: Modify for use with address/leak sanitizer

**Comment by levitte** [comment]
@richsalz, you've got a small update to do with test/memleaktest.c in #10572 :wink:

**Comment by richsalz** [comment]
yes, i rebased, took master for the conflicts, squashed, and pushed.  i'll note that over there too.

**Comment by mattcaswell** [comment]
This doesn't actually seem to work in travis. The memleak test is failing in the extended tests.

**Comment by mattcaswell** [comment]
I'm proposing to temporarily revert this in #10689.

**Pull request closed without merging**